streaming/src/main/scala/org/apache/spark/streaming/dstream/SocketInputDStream.scala

```diff
@@ -52,34 +52,40 @@ class SocketReceiver[T: ClassTag](
     storageLevel: StorageLevel
   ) extends Receiver[T](storageLevel) with Logging {
 
+  var socket: Socket = null
+
   def onStart() {
+    logInfo("Connecting to " + host + ":" + port)
+    socket = new Socket(host, port)
+    logInfo("Connected to " + host + ":" + port)
+
     // Start the thread that receives data over a connection
     new Thread("Socket Receiver") {
       setDaemon(true)
-      override def run() { receive() }
+      override def run() { receive(socket) }
     }.start()
   }
 
   def onStop() {
-    // There is nothing much to do as the thread calling receive()
-    // is designed to stop by itself isStopped() returns false
+    if (socket != null) {
+      socket.close()
```

Comment:
The finally block was left below though. It will cause the socket to be closed twice. It seems like it should be closed one place, or at least nulled in the finally block. But then you need to synchronize too.